From ab968ec4144d7527da27363c844394f1889f723d Mon Sep 17 00:00:00 2001
From: Zac Medico <zmedico@gentoo.org>
Date: Sun, 18 Apr 2021 16:17:16 -0700
Subject: [PATCH 1/2] Revert "Remove workarounds that are not needed with the
 current versions of sip-build"

This reverts commit 04ad7bc901c11ab559a7b33b2e6c304ad1d59116.
---
 setup/build.py | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/setup/build.py b/setup/build.py
index 6bd9b43452..d1c67dc522 100644
--- a/setup/build.py
+++ b/setup/build.py
@@ -524,6 +524,7 @@ def create_sip_build_skeleton(self, src_dir, ext):
 
 [tool.sip.project]
 sip-files-dir = "."
+sip-module = "PyQt5.sip"
 {abi_version}
 
 [tool.sip.bindings.pictureflow]
@@ -563,6 +564,15 @@ def build_pyqt_extension(self, ext, dest, sbf):
         cwd = os.getcwd()
         try:
             os.chdir(os.path.join(src_dir, 'build'))
+            if ext.needs_exceptions:
+                # bug in sip-build
+                for q in walk('.'):
+                    if os.path.basename(q) in ('Makefile',):
+                        with open(q, 'r+') as f:
+                            raw = f.read()
+                            raw = raw.replace('-fno-exceptions', '-fexceptions')
+                            f.seek(0), f.truncate()
+                            f.write(raw)
             self.check_call([self.env.make] + ([] if iswindows else ['-j%d'%(os.cpu_count() or 1)]))
             e = 'pyd' if iswindows else 'so'
             m = glob.glob(f'{ext.name}/{ext.name}.*{e}')
-- 
2.26.2

From 643263c558bb2fd0956a459b2a530b8ff69d779e Mon Sep 17 00:00:00 2001
From: Zac Medico <zmedico@gentoo.org>
Date: Sun, 18 Apr 2021 16:27:48 -0700
Subject: [PATCH 2/2] Revert "Move to SIP v5"

This reverts commit 7a4b3f61ff24f8c39c8d5cf86c54da9de9267025.
---
 bypy/linux/__main__.py     |   6 ++
 bypy/macos/__main__.py     |   3 +
 bypy/sources.json          |  57 +------------
 bypy/windows/__main__.py   |  82 ++++++++++++++++++-
 setup/build.py             | 160 +++++++++++++++++++++----------------
 setup/build_environment.py |  45 ++++++++++-
 setup/extensions.json      |   1 -
 7 files changed, 222 insertions(+), 132 deletions(-)

diff --git a/bypy/linux/__main__.py b/bypy/linux/__main__.py
index e0b07b1055..1a4778ab5a 100644
--- a/bypy/linux/__main__.py
+++ b/bypy/linux/__main__.py
@@ -143,6 +143,12 @@ def copy_python(env, ext_dir):
     dest = j(env.py_dir, 'site-packages')
     import_site_packages(srcdir, dest)
 
+    filter_pyqt = {x + '.so' for x in PYQT_MODULES} | {'sip.so'}
+    pyqt = j(dest, 'PyQt5')
+    for x in os.listdir(pyqt):
+        if x.endswith('.so') and x not in filter_pyqt:
+            os.remove(j(pyqt, x))
+
     for x in os.listdir(env.SRC):
         c = j(env.SRC, x)
         if os.path.exists(j(c, '__init__.py')):
diff --git a/bypy/macos/__main__.py b/bypy/macos/__main__.py
index 45ccae7aae..dc9d7d6099 100644
--- a/bypy/macos/__main__.py
+++ b/bypy/macos/__main__.py
@@ -554,6 +554,9 @@ def add_site_packages(self):
             if err.errno != errno.ENOENT:
                 raise
         sp = join(self.resources_dir, 'Python', 'site-packages')
+        for x in os.listdir(join(sp, 'PyQt5')):
+            if x.endswith('.so') and x.rpartition('.')[0] not in PYQT_MODULES and x != 'sip.so':
+                os.remove(join(sp, 'PyQt5', x))
         self.remove_bytecode(sp)
 
     @flush
diff --git a/bypy/sources.json b/bypy/sources.json
index 3767058255..9ed89314f6 100644
--- a/bypy/sources.json
+++ b/bypy/sources.json
@@ -815,63 +815,12 @@
         }
     },
 
-    {
-        "name": "toml",
-		"comment": "Needed for sip (build time dependency)",
-        "unix": {
-            "filename": "toml-0.10.1.tar.gz",
-            "hash": "sha256:926b612be1e5ce0634a2ca03470f95169cf16f939018233a670519cb4ac58b0f",
-            "urls": ["pypi"]
-        }
-    },
-
-    {
-        "name": "pyparsing",
-		"comment": "Needed for packaging (build time dependency)",
-        "unix": {
-            "filename": "pyparsing-2.4.7.tar.gz",
-            "hash": "sha256:c203ec8783bf771a155b207279b9bccb8dea02d8f0c9e5f8ead507bc3246ecc1",
-            "urls": ["pypi"]
-        }
-    },
-
-    {
-        "name": "packaging",
-		"comment": "Needed for sip (build time dependency)",
-        "unix": {
-            "filename": "packaging-20.4.tar.gz",
-            "hash": "sha256:4357f74f47b9c12db93624a82154e9b120fa8293699949152b22065d556079f8",
-            "urls": ["pypi"]
-        }
-    },
-
     {
         "name": "sip",
-		"comment": "build time dependency",
-        "unix": {
-            "filename": "sip-5.5.0.tar.gz",
-            "hash": "sha256:5d024c419b30fea8a6de8c71a560c7ab0bc3c221fbfb14d55a5b865bd58eaac5",
-            "urls": ["pypi"]
-        }
-    },
-
-    {
-        "name": "pyqt-builder",
-		"comment": "build time dependency",
         "unix": {
-            "filename": "PyQt-builder-1.6.0.tar.gz",
-            "hash": "sha256:fb80f01407718bfc68f78e276f2543fcfc5b5900b1187c1fd1f2ff51a6f2a13c",
-            "urls": ["pypi"]
-        }
-    },
-
-    {
-        "name": "pyqt-sip",
-		"comment": "runtime sip module for PyQt",
-        "unix": {
-            "filename": "PyQt5_sip-12.8.1.tar.gz",
-            "hash": "sha256:30e944db9abee9cc757aea16906d4198129558533eb7fadbe48c5da2bd18e0bd",
-            "urls": ["pypi"]
+            "filename": "sip-4.19.24.tar.gz",
+            "hash": "sha256:edcd3790bb01938191eef0f6117de0bf56d1136626c0ddb678f3a558d62e41e5",
+            "urls": ["https://www.riverbankcomputing.com/static/Downloads/sip/4.19.24/{filename}"]
         }
     },
 
diff --git a/bypy/windows/__main__.py b/bypy/windows/__main__.py
index 635feb0199..d9af16092b 100644
--- a/bypy/windows/__main__.py
+++ b/bypy/windows/__main__.py
@@ -156,8 +156,7 @@ def copybin(x):
     for x in glob.glob(os.path.join(env.python_base, 'DLLs', '*.dll')):  # dlls needed by python
         copybin(x)
     for f in walk(os.path.join(env.python_base, 'Lib')):
-        q = f.lower()
-        if q.endswith('.dll') and 'scintilla' not in q and 'pyqtbuild' not in q:
+        if f.lower().endswith('.dll') and 'scintilla' not in f.lower():
             copybin(f)
     ext_map = extract_extension_modules(ext_dir, env.dll_dir)
     ext_map.update(extract_extension_modules(j(env.python_base, 'DLLs'), env.dll_dir, move=False))
@@ -197,6 +196,11 @@ def ignore_lib(root, items):
     install_site_py(env)
     sp_dir = j(env.lib_dir, 'site-packages')
 
+    pyqt = j(env.lib_dir, 'site-packages', 'PyQt5')
+    for x in {x for x in os.listdir(pyqt) if x.endswith('.pyd')}:
+        if x.partition('.')[0] not in PYQT_MODULES and x != 'sip.pyd':
+            os.remove(j(pyqt, x))
+
     printf('Adding calibre sources...')
     for x in glob.glob(j(CALIBRE_DIR, 'src', '*')):
         if os.path.isdir(x):
@@ -514,6 +518,80 @@ def build_launchers(env, incdir, debug=False):
             run(*cmd)
 
 
+def add_to_zipfile(zf, name, base, zf_names):
+    if '__pycache__' in name:
+        return
+    abspath = j(base, name)
+    name = name.replace(os.sep, '/')
+    if name in zf_names:
+        raise ValueError('Already added %r to zipfile [%r]' % (name, abspath))
+    zinfo = zipfile.ZipInfo(filename=name, date_time=(1980, 1, 1, 0, 0, 0))
+
+    if os.path.isdir(abspath):
+        if not os.listdir(abspath):
+            return
+        zinfo.external_attr = 0o700 << 16
+        zf.writestr(zinfo, '')
+        for x in os.listdir(abspath):
+            add_to_zipfile(zf, name + os.sep + x, base, zf_names)
+    else:
+        ext = os.path.splitext(name)[1].lower()
+        if ext in ('.dll',):
+            raise ValueError('Cannot add %r to zipfile' % abspath)
+        zinfo.external_attr = 0o600 << 16
+        if ext in ('.py', '.pyc', '.pyo'):
+            with open(abspath, 'rb') as f:
+                zf.writestr(zinfo, f.read())
+
+    zf_names.add(name)
+
+
+def archive_lib_dir(env):
+    printf('Putting all python code into a zip file for performance')
+    zf_names = set()
+    with zipfile.ZipFile(env.pylib, 'w', zipfile.ZIP_STORED) as zf:
+        # Add everything in Lib except site-packages to the zip file
+        for x in os.listdir(env.lib_dir):
+            if x == 'site-packages':
+                continue
+            add_to_zipfile(zf, x, env.lib_dir, zf_names)
+
+        sp = j(env.lib_dir, 'site-packages')
+        # Special handling for pywin32
+        handled = {'pywin32.pth', 'win32'}
+        base = j(sp, 'win32', 'lib')
+        for x in os.listdir(base):
+            if os.path.splitext(x)[1] not in ('.exe',) and x != '__pycache__':
+                add_to_zipfile(zf, x, base, zf_names)
+        base = os.path.dirname(base)
+        for x in os.listdir(base):
+            if not os.path.isdir(j(base, x)):
+                if os.path.splitext(x)[1] not in ('.exe',):
+                    add_to_zipfile(zf, x, base, zf_names)
+
+        # We dont want the site.py (if any) from site-packages
+        handled.add('site.pyo')
+        handled.add('site.pyc')
+        handled.add('site.py')
+        handled.add('sitecustomize.pyo')
+        handled.add('sitecustomize.pyc')
+        handled.add('sitecustomize.py')
+
+        # The rest of site-packages
+        for x in os.listdir(sp):
+            if x in handled or x.endswith('.egg-info'):
+                continue
+            absp = j(sp, x)
+            if os.path.isdir(absp):
+                if not os.listdir(absp):
+                    continue
+                add_to_zipfile(zf, x, sp, zf_names)
+            else:
+                add_to_zipfile(zf, x, sp, zf_names)
+
+    shutil.rmtree(env.lib_dir)
+
+
 def copy_crt_and_d3d(env):
     printf('Copying CRT and D3D...')
     plat = ('x64' if is64bit else 'x86')
diff --git a/setup/build.py b/setup/build.py
index d1c67dc522..3fd580498f 100644
--- a/setup/build.py
+++ b/setup/build.py
@@ -8,7 +8,7 @@
 import textwrap, os, shlex, subprocess, glob, shutil, sys, json
 from collections import namedtuple
 
-from setup import Command, islinux, isbsd, isfreebsd, ismacos, ishaiku, SRC, iswindows
+from setup import Command, islinux, isbsd, isfreebsd, ismacos, ishaiku, SRC, iswindows, __version__
 isunix = islinux or ismacos or isbsd or ishaiku
 
 py_lib = os.path.join(sys.prefix, 'libs', 'python%d%d.lib' % sys.version_info[:2])
@@ -16,12 +16,6 @@
 LinkCommand = namedtuple('LinkCommand', 'cmd objects dest')
 
 
-def walk(path='.'):
-    for dirpath, dirnames, filenames in os.walk(path):
-        for f in filenames:
-            yield os.path.join(dirpath, f)
-
-
 def init_symbol_name(name):
     prefix = 'PyInit_'
     return prefix + name
@@ -41,7 +35,6 @@ def __init__(self, name, sources, **kwargs):
         self.needs_py2 = d['needs_py2'] = kwargs.get('needs_py2', False)
         self.headers = d['headers'] = absolutize(kwargs.get('headers', []))
         self.sip_files = d['sip_files'] = absolutize(kwargs.get('sip_files', []))
-        self.needs_exceptions = d['needs_exceptions'] = kwargs.get('needs_exceptions', False)
         self.inc_dirs = d['inc_dirs'] = absolutize(kwargs.get('inc_dirs', []))
         self.lib_dirs = d['lib_dirs'] = absolutize(kwargs.get('lib_dirs', []))
         self.extra_objs = d['extra_objs'] = absolutize(kwargs.get('extra_objs', []))
@@ -78,6 +71,7 @@ def __init__(self, name, sources, **kwargs):
             flag = '/O%d' if iswindows else '-O%d'
             of = flag % of
         self.cflags.insert(0, of)
+        self.qt_private_headers = d['qt_private_headers'] = kwargs.get('qt_private', [])
 
 
 def lazy_load(name):
@@ -261,6 +255,9 @@ class Build(Command):
            PODOFO_LIB_DIR - podofo library files
 
            QMAKE          - Path to qmake
+           SIP_BIN        - Path to the sip binary
+           VS90COMNTOOLS  - Location of Microsoft Visual Studio 9 Tools (windows only)
+
         ''')
 
     def add_options(self, parser):
@@ -343,8 +340,7 @@ def run(self, opts):
                 raise SystemExit(1)
         for (ext, dest) in pyqt_extensions:
             sbf = sbf_map[id(ext)]
-            if not os.path.exists(sbf):
-                self.build_pyqt_extension(ext, dest, sbf)
+            self.build_pyqt_extension(ext, dest, sbf)
 
         if opts.only in {'all', 'headless'}:
             self.build_headless()
@@ -502,75 +502,101 @@ def build_headless(self):
         if ismacos:
             os.rename(self.j(self.d(target), 'libheadless.dylib'), self.j(self.d(target), 'headless.so'))
 
-    def create_sip_build_skeleton(self, src_dir, ext):
-        from setup.build_environment import pyqt_sip_abi_version
-        abi_version = ''
-        if pyqt_sip_abi_version():
-            abi_version = f'abi-version = "{pyqt_sip_abi_version()}"'
-        sipf = ext.sip_files[0]
-        needs_exceptions = 'true' if ext.needs_exceptions else 'false'
-        with open(os.path.join(src_dir, 'pyproject.toml'), 'w') as f:
-            f.write(f'''
-[build-system]
-requires = ["sip >=5.3", "PyQt-builder >=1"]
-build-backend = "sipbuild.api"
-
-[tool.sip.metadata]
-name = "{ext.name}"
-requires-dist = "PyQt5 (>=5.15)"
-
-[tool.sip]
-project-factory = "pyqtbuild:PyQtProject"
-
-[tool.sip.project]
-sip-files-dir = "."
-{abi_version}
-
-[tool.sip.bindings.pictureflow]
-headers = {ext.headers}
-sources = {ext.sources}
-exceptions = {needs_exceptions}
-include-dirs = {ext.inc_dirs}
-qmake-QT = ["widgets"]
-sip-file = "{os.path.basename(sipf)}"
-''')
-        shutil.copy2(sipf, src_dir)
-
     def get_sip_commands(self, ext):
-        from setup.build_environment import QMAKE
         pyqt_dir = self.j(self.build_dir, 'pyqt')
         src_dir = self.j(pyqt_dir, ext.name)
-        # TODO: Handle building extensions with multiple SIP files.
-        sipf = ext.sip_files[0]
+        from setup.build_environment import pyqt
+        sip_files = ext.sip_files
+        ext.sip_files = []
+        sipf = sip_files[0]
+        os.makedirs(src_dir, exist_ok=True)
         sbf = self.j(src_dir, self.b(sipf)+'.sbf')
         cmd = None
         cwd = None
-        if self.newer(sbf, [sipf] + ext.headers + ext.sources):
-            shutil.rmtree(src_dir, ignore_errors=True)
-            os.makedirs(src_dir)
-            self.create_sip_build_skeleton(src_dir, ext)
+        if self.newer(sbf, [sipf]+ext.headers):
+            shutil.rmtree(src_dir)
+            os.mkdir(src_dir)
             cwd = src_dir
-            cmd = [
-                sys.executable, '-c',
-                '''from sipbuild.tools.build import main; main();''',
-                '--verbose', '--no-make', '--qmake', QMAKE
-            ]
+            cmd = [pyqt['sip_bin'], '-w', '-c', src_dir, '-I' + pyqt['pyqt_sip_dir']] + shlex.split(pyqt['sip_flags']) + [sipf]
         return cmd, sbf, cwd
 
+    def get_sip_data(self, sbf):
+        if os.path.exists(sbf):
+            with open(sbf) as f:
+                return json.loads(f.read())
+        src_dir = os.path.dirname(sbf)
+
+        def transform(x):
+            return x.replace(os.sep, '/')
+
+        ans = {
+            'target': os.path.basename(src_dir),
+            'sources': list(map(transform, glob.glob(os.path.join(src_dir, '*.cpp')))),
+            'headers': list(map(transform, glob.glob(os.path.join(src_dir, '*.h')))),
+        }
+        with open(sbf, 'w') as f:
+            f.write(json.dumps(ans))
+        return ans
+
     def build_pyqt_extension(self, ext, dest, sbf):
         self.info(f'\n####### Building {ext.name} extension', '#'*7)
-        src_dir = os.path.dirname(sbf)
+        from setup.build_environment import pyqt, qmakespec, QMAKE
+        from setup.parallel_build import cpu_count
+        from distutils import sysconfig
+        pyqt_dir = self.j(self.build_dir, 'pyqt')
+        src_dir = self.j(pyqt_dir, ext.name)
+        if not os.path.exists(src_dir):
+            os.makedirs(src_dir)
+        sip = self.get_sip_data(sbf)
+        pro = textwrap.dedent(
+        '''\
+        TEMPLATE = lib
+        CONFIG += release plugin
+        QT += widgets
+        TARGET = {target}
+        HEADERS = {headers}
+        SOURCES = {sources}
+        INCLUDEPATH += {sipinc} {pyinc}
+        VERSION = {ver}
+        win32 {{
+            LIBS += {py_lib}
+            TARGET_EXT = .dll
+        }}
+        macx {{
+            QMAKE_LFLAGS += "-undefined dynamic_lookup"
+        }}
+        ''').format(
+            target=sip['target'], headers=' '.join(sip['headers'] + ext.headers), sources=' '.join(ext.sources + sip['sources']),
+            sipinc=pyqt['sip_inc_dir'], pyinc=sysconfig.get_python_inc(), py_lib=py_lib,
+            ver=__version__
+        )
+        for incdir in ext.inc_dirs:
+            pro += '\nINCLUDEPATH += ' + incdir
+        if not iswindows and not ismacos:
+            # Ensure that only the init symbol is exported
+            pro += '\nQMAKE_LFLAGS += -Wl,--version-script=%s.exp' % sip['target']
+            with open(os.path.join(src_dir, sip['target'] + '.exp'), 'wb') as f:
+                f.write(('{ global: %s; local: *; };' % init_symbol_name(sip['target'])).encode('utf-8'))
+        if ext.qt_private_headers:
+            qph = ' '.join(x + '-private' for x in ext.qt_private_headers)
+            pro += '\nQT += ' + qph
+        proname = '%s.pro' % sip['target']
+        with open(os.path.join(src_dir, proname), 'wb') as f:
+            f.write(pro.encode('utf-8'))
         cwd = os.getcwd()
+        qmc = []
+        if iswindows:
+            qmc += ['-spec', qmakespec]
+        fext = 'dll' if iswindows else 'dylib' if ismacos else 'so'
+        name = '%s%s.%s' % ('release/' if iswindows else 'lib', sip['target'], fext)
         try:
-            os.chdir(os.path.join(src_dir, 'build'))
-            self.check_call([self.env.make] + ([] if iswindows else ['-j%d'%(os.cpu_count() or 1)]))
-            e = 'pyd' if iswindows else 'so'
-            m = glob.glob(f'{ext.name}/{ext.name}.*{e}')
-            if len(m) != 1:
-                raise SystemExit(f'Found extra PyQt extension files: {m}')
-            shutil.copy2(m[0], dest)
-            with open(sbf, 'w') as f:
-                f.write('done')
+            os.chdir(src_dir)
+            if self.newer(dest, sip['headers'] + sip['sources'] + ext.sources + ext.headers):
+                self.check_call([QMAKE] + qmc + [proname])
+                self.check_call([self.env.make]+([] if iswindows else ['-j%d'%(cpu_count or 1)]))
+                shutil.copy2(os.path.realpath(name), dest)
+                if iswindows and os.path.exists(name + '.manifest'):
+                    shutil.copy2(name + '.manifest', dest + '.manifest')
         finally:
             os.chdir(cwd)
 
diff --git a/setup/build_environment.py b/setup/build_environment.py
index 41dccb1021..ef4207f4ae 100644
--- a/setup/build_environment.py
+++ b/setup/build_environment.py
@@ -6,10 +6,10 @@
 __copyright__ = '2009, Kovid Goyal <kovid@kovidgoyal.net>'
 __docformat__ = 'restructuredtext en'

-import os, subprocess, re, shutil
+import os, subprocess, re, shutil, sys, sysconfig
 from functools import lru_cache

-from setup import ismacos, iswindows, is64bit, islinux, ishaiku
+from setup import ismacos, iswindows, is64bit, islinux, isfreebsd, ishaiku

 NMAKE = RC = msvc = MT = win_inc = win_lib = win_cc = win_ld = None

@@ -45,8 +45,6 @@ def merge_paths(a, b):
         QMAKE = q
         break
 QMAKE = os.environ.get('QMAKE', QMAKE)
-if iswindows and not QMAKE.lower().endswith('.exe'):
-    QMAKE += '.exe'
 
 PKGCONFIG = shutil.which('pkg-config')
 PKGCONFIG = os.environ.get('PKG_CONFIG', PKGCONFIG)
@@ -98,8 +96,47 @@ def readvar(name):
     return re.search('^%s:(.+)$' % name, qraw, flags=re.M).group(1).strip()
 
 
+pyqt = {x:readvar(y) for x, y in (
+    ('inc', 'QT_INSTALL_HEADERS'), ('lib', 'QT_INSTALL_LIBS')
+)}
 qt = {x:readvar(y) for x, y in {'libs':'QT_INSTALL_LIBS', 'plugins':'QT_INSTALL_PLUGINS'}.items()}
 qmakespec = readvar('QMAKE_SPEC') if iswindows else None
+
+pyqt['sip_bin'] = os.environ.get('SIP_BIN', 'sip')
+
+import PyQt5
+from PyQt5.QtCore import PYQT_CONFIGURATION
+pyqt['sip_flags'] = PYQT_CONFIGURATION['sip_flags']
+
+
+def get_sip_dir():
+    q = None
+    if getattr(PyQt5, '__file__', None):
+        q = os.path.join(os.path.dirname(PyQt5.__file__), 'bindings')
+        if not os.path.exists(q):
+            q = None
+    if q is None:
+        if iswindows:
+            q = os.path.join(sys.prefix, 'share', 'sip')
+        elif isfreebsd:
+            q = os.path.join(sys.prefix, 'share', 'py-sip')
+        else:
+            q = os.path.join(os.path.dirname(PyQt5.__file__), 'bindings')
+            if not os.path.exists(q):
+                q = os.path.join(sys.prefix, 'share', 'sip')
+    q = os.environ.get('SIP_DIR', q)
+    for x in ('', 'Py2-PyQt5', 'PyQt5', 'sip/PyQt5'):
+        base = os.path.join(q, x)
+        if os.path.exists(os.path.join(base, 'QtWidgets')):
+            return base
+    raise EnvironmentError('Failed to find the location of the PyQt5 .sip files')
+
+
+pyqt['pyqt_sip_dir'] = get_sip_dir()
+pyqt['sip_inc_dir'] = os.environ.get('SIP_INC_DIR', sysconfig.get_path('include'))
+
+qt_inc = pyqt['inc']
+qt_lib = pyqt['lib']
 ft_lib_dirs = []
 ft_libs = []
 ft_inc_dirs = []
diff --git a/setup/extensions.json b/setup/extensions.json
index 4900ff5b88..7e8cd8e42f 100644
--- a/setup/extensions.json
+++ b/setup/extensions.json
@@ -139,7 +139,6 @@
         "sources": "calibre/utils/imageops/imageops.cpp calibre/utils/imageops/quantize.cpp calibre/utils/imageops/ordered_dither.cpp",
         "headers": "calibre/utils/imageops/imageops.h",
         "sip_files": "calibre/utils/imageops/imageops.sip",
-		"needs_exceptions": true,
         "inc_dirs": "calibre/utils/imageops"
     },
     {
-- 
2.26.2

